sparkplug-core/src/java/sparkplug/function/SerializableFn.java

```diff
@@ -28,33 +45,99 @@ private SerializableFn() {
+            out.writeInt(namespaces.size());
+            for (String ns : namespaces) {
+                out.writeObject(ns);
+            }
+            // Write out the function itself.
+            out.writeObject(f);
+        } catch (IOException ex) {
+            logger.error("Error serializing function " + f, ex);
+            throw ex;
+        } catch (RuntimeException ex){
+            logger.error("Error serializing function " + f, ex);
+            throw ex;
+        }
     }
 
 
+    /**
+     * Deserialize a function from the provided input stream.
+     * An unspoken part of the `Serializable` interface.
+     *
+     * @param in stream to read the function from
+     */
     private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
-        String namespace = (String)in.readObject();
-        Symbol namespaceSym = Symbol.intern(namespace);
-        // TODO: Try optimizing by doing an unsynchronized resolve first.
-        synchronized (RT.REQUIRE_LOCK) {
-            RT.var("clojure.core", "require").invoke(namespaceSym);
+        String className = "";
+        try {
+            // Read the function class name.
+            className = (String)in.readObject();
+            logger.trace("Deserializing " + className);
+            // Read the referenced namespaces and load them.
+            int nsCount = in.readInt();
+            this.namespaces = new ArrayList<String>(nsCount);
+            for (int i = 0; i < nsCount; i++) {
+                String ns = (String)in.readObject();
+                namespaces.add(ns);
+                requireNamespace(ns);
+            }
+            // Read the function itself.
+            this.f = (IFn)in.readObject();
+        } catch (IOException ex) {
+            logger.error("IO error deserializing function " + className, ex);
+            throw ex;
+        } catch (ClassNotFoundException ex) {
+            logger.error("Class error deserializing function " + className, ex);
+            throw ex;
+        } catch (RuntimeException ex) {
+            logger.error("Error deserializing function " + className, ex);
+            throw ex;
         }
-        this.f = (IFn)in.readObject();
     }
 
 
     /**
-     * Construct a new serializable wrapper for the function with an explicit
-     * set of required namespaces.
+     * Load the namespace specified by the given symbol.
      *
-     * @param fn Clojure function to wrap
+     * @param namespace string designating the namespace to load
      */
-    protected SerializableFn(IFn fn) {
-        this.f = fn;
+    private static void requireNamespace(String namespace) {
+        try {
+            logger.trace("(require " + namespace + ")");
+            synchronized (RT.REQUIRE_LOCK) {
+                Symbol sym = Symbol.intern(namespace);
+                require.invoke(sym);
+            }
+        } catch (Exception ex) {
+            logger.warn("Error loading namespace " + namespace, ex);
```

Comment:
Seems like a pretty serious error, should it be rethrown instead of swallowed?

Reply:
I could go both ways - if this require fails but the function succeeds on invocation, is that ok?